File: CPlusPlus/Sample_Source/Text/ExtractTextFromAnnotations/ExtractTextFromAnnotations.cpp
//
// Copyright (c) 2023, Datalogics, Inc. All rights reserved.
//
// For complete copyright information, refer to:
// http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
//
//  This sample extracts text from the Annotations in a PDF
//  document and saves the text to a file.
//
// For more detail see the description of the ExtractTextFromAnnotations sample program on our Developer’s
// site, http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/c1samples

#include <fstream>
#include <string>

#include "InitializeLibrary.h"
#include "APDFLDoc.h"
#include "TextExtract.h"

#ifdef MAC_PLATFORM
#include "../../_Common/ThirdParty/json.hpp"
#else
#include "ThirdParty/json.hpp"
#endif

// Set the namespace for the JSON library
using json = nlohmann::json;

const char *DEF_INPUT = "../../../../Resources/Sample_Input/sample_annotations.pdf";
const char *DEF_OUTPUT = "ExtractTextFromAnnotations-out.json";

int main(int argc, char **argv) {

    // Initialize the library
    APDFLib libInit;
    ASErrorCode errCode = 0;
    if (libInit.isValid() == false) {
        errCode = libInit.getInitError();
        std::cout << "Initialization failed with code " << errCode << std::endl;
        return libInit.getInitError();
    }

    DURING

        APDFLDoc inAPDoc(DEF_INPUT, true);

        // This array will hold the JSON stream that we will print to the output JSON file.
        json result = json::array();

        // Create the TextExtract object
        TextExtract textExtract(inAPDoc.getPDDoc());

        std::vector<PDAnnotationExtractRec> extractedText = textExtract.GetAnnotationText();

        for (size_t textIndex = 0; textIndex < extractedText.size(); ++textIndex)
        {
            json textObject = json::object();

            textObject["annotation-type"] = extractedText[textIndex].type;
            textObject["annotation-text"] = extractedText[textIndex].text;
            result.push_back(textObject);
        }
        std::ofstream jsonOutputFileStream(DEF_OUTPUT, std::ios_base::trunc | std::ios_base::out);
        jsonOutputFileStream << std::setw(4) << result;
        jsonOutputFileStream.close();

    HANDLER
        errCode = ERRORCODE;
        libInit.displayError(errCode);
    END_HANDLER

    return errCode;
}

File: DotNET/Sample_Source/Text/ExtractTextFromAnnotations/ExtractTextFromAnnotations.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using Datalogics.PDFL;
using ExtractTextNameSpace;

/*
 * 
 * This sample extracts text from the Annotations in a PDF
 * document and saves the text to a file.
 *
 * Copyright (c) 2022-2023, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/.
 *
 */
namespace ExtractTextFromAnnotations
{
    class ExtractTextFromAnnotations
    {

        // Set Defaults
        static String sInput = Library.ResourceDirectory + "Sample_Input/sample_annotations.pdf";
        static String sOutput = "../ExtractTextFromAnnotations-out.json";

        static void Print(AnnotationTextObject t)
        {
            Console.WriteLine("Annotation Type > " + t.AnnotationType);
            Console.WriteLine("Annotation Text > " + t.AnnotationText);
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Annotations Sample:");


            using (new Library())
            {
                Console.WriteLine("Initialized the library.");

                String sInput = Library.ResourceDirectory + "Sample_Input/sample_annotations.pdf";

                Console.WriteLine("Input file: " + sInput);

                using (Document doc = new Document(sInput))
                {
                    using (ExtractText docAnnotText = new ExtractText(doc))
                    {
                        List<AnnotationTextObject> result = docAnnotText.GetAnnotationText();

                        // Print the output to the console
                        result.ForEach(Print);

                        // Save the output to a JSON file.
                        Console.WriteLine("Writing JSON to " + sOutput);
                        JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };
                        string json = JsonSerializer.Serialize(result, options);
                        System.IO.File.WriteAllText(sOutput, json);
                    }
                }
            }
        }
    }
}
